### PR TITLE
cosmos: refresh topology on 403/1008 + 403/3

### PR DESCRIPTION
**Pull request by tvaron3**
### Retry semantics
- `try_handle_database_account_not_found` (1008) and `try_handle_write_forbidden` (403/3) now emit `LocationEffect::RefreshAccountProperties` and consume the dedicated `backend_failover_retry_count` budget (120 attempts × 1 s pacing) instead of bubbling up.
- The original 4xx substatus is surfaced unchanged when the budget exhausts (not rewritten as 503) — preserves the diagnostic signal.
- `try_handle_write_forbidden` suppresses `MarkEndpointUnavailable` when `is_ppcb_managed` is true (multi-write + partitioned + PPCB active); per-partition mark and topology refresh still fire so other partitions on the same endpoint keep writing normally.

### Per-partition routing
- `ppcb_should_skip` in `resolve_endpoint` treats `OperationOptions::excluded_regions` as a hard filter on the PPCB override path; previously the override fast-path could silently route to a region the caller had excluded.
- `ppcb_should_skip` also skips when the override's `current_endpoint` no longer appears in the post-refresh topology (the override map is never pruned by `RefreshAccountProperties`).
- `evaluate_transport_layer_outcome`'s `definitely_not_sent` branch no longer emits `MarkPartitionUnavailable` — gateway-mode connect failures stop inflating per-partition PPCB counters.

### State simplification
- Dropped `OperationRetryState::attempted_failed_regions` and the `failed_region` parameter on `advance_to_next_attempt`. `LocationIndex` advancement via `next_for_generation` already rotates past the just-failed region for the multi-write PPCB-managed path; the previous field was a parallel mechanism whose unique invariant was unobservable in production code paths. Documented edge case: after a generation bump the rebased index can retry the just-failed region exactly once, bounded by the failover budget.

### Tests
- `tests/in_memory_emulator_tests/topology_refresh_on_substatus.rs` — new file. Covers 1008 across read/write/query/feed-range, all-regions-stale, after-refresh-still-1008, 403/3 stale-write-region, and 403/3 all-regions-forbidden.
- New PPCB override skip-reason tests in `operation_pipeline.rs` for endpoint-marked, excluded-region, and stale-topology cases.
- Multi-write integration coverage in `cosmos_multi_write_fault_injection.rs`.

## Verification

- `cargo build --tests --all-features` — clean.
- `cargo clippy --tests --all-features` — clean.
- `cargo test --lib` — 1780 / 0.
- In-memory emulator integration tests in `topology_refresh_on_substatus.rs` — all pass.
- Live drill reproduction in centralus blackhole (image `drill-upstream-5e850f0c9`) — the original 47k × 1008 / 8k × 403-3 cascade no longer bubbles up; the SDK refreshes and converges on `eastus2`.

### CODE REVIEW

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/retry_evaluation.rs [hunk 1]
```diff
@@ -384,29 +397,81 @@ fn try_handle_write_forbidden(
...
+        (
+            retry_state.clone().advance_backend_failover(),
+            Some(BACKEND_FAILOVER_RETRY_INTERVAL),
+        )
+    } else {
+        if !retry_state.can_retry_failover() {
+            return None;
+        }
+        (retry_state.clone().advance_failover(), None)
+    };
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/retry_evaluation.rs:418]: The single-master 403/3 path is staying on the generic 4-attempt budget for now and will be revisited during the upcoming single-master DR drill. Because single-master can fall back to both regions regardless of `excluded_regions`, the 4 attempts may already cover the convergence window — we'll confirm or change based on drill data. The docs/changelog are accurate for 403/3 single-write (3 retries × 0 ms generic); only 1008 was unified and that fix is landing in the other thread.

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md [hunk 2]
```diff
@@ -49,13 +49,21 @@ These are deterministic client errors. No retry will change the outcome.
...
-| Substatus | Meaning | Action | Budget |
-|-----------|---------|--------|--------|
-| 3 | Write Forbidden (region failover) | Cross-region failover retry | 3 failover attempts |
-| 1008 | Write Forbidden (partition moved) | Cross-region failover retry | 3 failover attempts |
-| Other | Permission denied | Abort | — |
+| Substatus | Meaning | Action | Budget (multi-write) | Budget (single-write) |
+|-----------|---------|--------|----------------------|-----------------------|
+| 3 | `WriteForbidden` — region is not currently a valid write region for this partition (writes only) | Refresh account topology + cross-region failover retry | **120** attempts × **1000 ms** delay (dedicated `backend_failover_retry_count`) | 3 attempts × 0 ms delay (shared generic budget) |
+| 1008 | `DatabaseAccountNotFound` — region no longer owns this account (all op types, including reads, writes, queries, feed-range queries, metadata) | Refresh account topology + cross-region failover retry | **120** attempts × **1000 ms** delay (dedicated `backend_failover_retry_count`) | 3 attempts × 0 ms delay (shared generic budget) |
+| Other | Permission denied | Abort | — | — |
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:56]: Fixed — 1008 was simplified in `ccae935d0` to use the backend-failover budget for both single-write and multi-write. Updated the table. The 403/3 row is correct as written: single-write 403/3 keeps the generic budget; only 1008 was unified.

File: sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,11 @@
...
+- `evaluate_transport_layer_outcome`'s `definitely_not_sent` branch now emits only `MarkEndpointUnavailable` (the `MarkPartitionUnavailable` emit was dropped) so Gateway-mode connect failures stop inflating per-partition PPCB counters. ([#4590](https://github.com/Azure/azure-sdk-for-rust/pull/4590))
+- `try_handle_write_forbidden` now suppresses `MarkEndpointUnavailable` when `is_ppcb_managed` is true (multi-write + partitioned + PPCB active); the per-partition mark and topology refresh still fire. ([#4590](https://github.com/Azure/azure-sdk-for-rust/pull/4590))
+- `ppcb_should_skip` in `resolve_endpoint` now treats `OperationOptions::excluded_regions` as a hard filter on the PPCB override path; previously the override fast-path bypassed the exclusion check enforced in `try_select_endpoint`. ([#4590](https://github.com/Azure/azure-sdk-for-rust/pull/4590))
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md]: Updated — the bullet conflated the two. 1008 uses the backend-failover budget for both single-master and multi-master (collapsed in `ccae935d0`). 403/3 only uses it on multi-master; single-master 403/3 stays on the generic budget pending the upcoming single-master DR drill.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/operation_pipeline.rs [hunk 4]
```diff
@@ -5171,6 +5242,710 @@ mod tests {
...
+            Duration::from_secs(60),
+        );
+        assert_ne!(
+            routing.endpoint, central,
+            "BUG: PPCB override pinned the partition to centralus even though \
+             centralus is in account.unavailable_endpoints. This is the \
+             production 4-attempt-all-to-central failure mode. The override-skip \
+             check at operation_pipeline.rs:1083 must also consult \
+             account.unavailable_endpoints (and excluded_regions); see sibling \
+             branch commit a7819f73 for the canonical fix."
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/operation_pipeline.rs]: Done — replaced the brittle `operation_pipeline.rs:1083` + commit-hash reference with an invariant-based message that names the contract (PPCB override path must honor `account.unavailable_endpoints` and `excluded_regions` the same way `try_select_endpoint` does).

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md [hunk 5]
```diff
@@ -211,7 +219,9 @@ Without PPCB, the driver marks entire endpoints as unavailable when errors occur
 | Layer | Budget | Scope |
 |-------|--------|-------|
 | Transport (429) | 9 attempts or 30s | Per-request, local only |
-| Operation failover | 3 attempts | Per-operation, cross-region |
+| Operation failover (generic — 5xx, 408, 410, transport) | 3 attempts | Per-operation, cross-region |
+| Backend-failover (403/3, 403/1008) — multi-write | **120 attempts × 1000 ms** | Per-operation, cross-region |
+| Backend-failover (403/3, 403/1008) — single-write | 3 attempts × 0 ms (generic) | Per-operation, cross-region |
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md]: Updated — split into three rows: 1008 backend-failover (single + multi), 403/3 multi-write backend-failover, 403/3 single-write generic. Now matches the code and the per-status table above.

File: sdk/cosmos/azure_data_cosmos_driver/tests/in_memory_emulator_tests/regional_gateway_unreachable.rs [hunk 6]
```diff
@@ -0,0 +1,774 @@
...
+) -> Fixture {
+    if enable_ppcb {
+        std::env::set_var("AZURE_COSMOS_PER_PARTITION_CIRCUIT_BREAKER_ENABLED", "true");
+        std::env::set_var("AZURE_COSMOS_CIRCUIT_BREAKER_FAILURE_COUNT_FOR_READS", "1");
+        std::env::set_var("AZURE_COSMOS_CIRCUIT_BREAKER_FAILURE_COUNT_FOR_WRITES", "1");
+    } else {
+        std::env::remove_var("AZURE_COSMOS_PER_PARTITION_CIRCUIT_BREAKER_ENABLED");
+        std::env::remove_var("AZURE_COSMOS_CIRCUIT_BREAKER_FAILURE_COUNT_FOR_READS");
+        std::env::remove_var("AZURE_COSMOS_CIRCUIT_BREAKER_FAILURE_COUNT_FOR_WRITES");
+    }
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/tests/in_memory_emulator_tests/regional_gateway_unreachable.rs]: Done — removed the env-var mutation from `build_fixture` and dropped the `enable_ppcb: bool` parameter. PPCB is already configured per-test through `ppcb_options()`, which sets `OperationOptions::per_partition_circuit_breaker_enabled` and the two threshold fields directly (per-instance, no process-global state). All call sites updated; the 15 `regional_gateway_unreachable` tests still pass.

**Review by analogrelay — APPROVED** [review]
Looks good. To confirm, we _do_ ignore excluded regions on metadata refresh operations, right?

**Review by analogrelay — APPROVED** [review]

**Review by simorenoh — APPROVED** [review]

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md [hunk 7]
```diff
@@ -49,13 +49,21 @@ These are deterministic client errors. No retry will change the outcome.
...
 
-| Substatus | Meaning | Action | Budget |
-|-----------|---------|--------|--------|
-| 3 | Write Forbidden (region failover) | Cross-region failover retry | 3 failover attempts |
-| 1008 | Write Forbidden (partition moved) | Cross-region failover retry | 3 failover attempts |
-| Other | Permission denied | Abort | — |
+| Substatus | Meaning | Action | Budget (multi-write) | Budget (single-write) |
+|-----------|---------|--------|----------------------|-----------------------|
+| 3 | `WriteForbidden` — region is not currently a valid write region for this partition (writes only) | Refresh account topology + cross-region failover retry | **120** attempts × **1000 ms** delay (dedicated `backend_failover_retry_count`) | 3 attempts × 0 ms delay (shared generic budget) |
+| 1008 | `DatabaseAccountNotFound` — region no longer owns this account (all op types, including reads, writes, queries, feed-range queries, metadata) | Refresh account topology + cross-region failover retry | **120** attempts × **1000 ms** delay (dedicated `backend_failover_retry_count`) | **120** attempts × **1000 ms** delay (dedicated `backend_failover_retry_count`) |
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:55]: In general static back-offs (like 1 second here) are a bad idea - when there is sustained impact it result sin too noisy retries (which could make things worse when impact is caused by resource contention on the client or server). All our retry policies in rust should use exponential backoffs. Also, 120 retries ver 2 minutes seems excessive - if you want to keep 2 minutes that is fine with me thought as long as you use exponential backoffs.

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md [hunk 8]
```diff
@@ -49,13 +49,21 @@ These are deterministic client errors. No retry will change the outcome.
...
+
+**Why a dedicated 120-attempt × 1s budget on multi-write?** The 3-attempt generic failover budget is exhausted long before the backend's topology change finishes propagating, turning a recoverable convergence window into a hard application-visible failure. 120 attempts × 1 s ≈ 2 minutes of bounded retries gives the backend time to settle while still guaranteeing eventual bubble-up. Without the delay, the SDK hot-loops 120 cross-region requests in well under a second — faster than the backend convergence window it is reacting to. Exponential backoff was rejected because the bound is already short, the signal is a topology change rather than a load spike, and growth would push worst-case wall time into the 10-minute range.
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:60]: >Exponential backoff was rejected because the bound is already short, the signal is a topology change rather than a load spike, and growth would push worst-case wall time into the 10-minute range.

This seems off - it would only push it into 10-minutes range when you keep 120 retries - which is not necessary - effectively you are saying you want to not bubble up failures to app as long as the topology update finishes within 2 minutes - exponential backoffs over max. 2 minutes would do the same. Also the vast majority of client workloads will perceive 2 minutes latency as availability loss anyway. I think in this specific case not using an exponential back-off would be acceptable only because you also limit the actual I/OP calls for account topology update to 1 every 5 seconds. But IMO an exponential backoff - strating with maybe 200ms and having an upper limti of somewehre around2-3 seconds back-off would make more sense. When topology update finsihes reaosnably fast it limits the impact on latency of the then successful responses. But as long as the 1 I/O operation per 5 seconds is enforced I would not block on this.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - except that I don't buy the argument why using 1s static back-off vs. exponential back-off. But given the actual I/O operations are limited to 1 / 5s - I am not blocking on it.

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md (hunk 8 above)
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:60]: I will followup with another pr to iterate on this change to add exponential back off. Wondering if we should add exponential back off for other failures as well. I am also running tests to see how long it takes 1 partition to stop throwing 403.3s based on data size.

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md (hunk 7 above)
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:55]: will followup in next pr

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md (hunk 8 above)
Comment (analogrelay) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:60]: @tvaron3 Can you create a quick issue to track that work, so I can get it on the backlog? We can include reviewing the retry interval entirely because I do agree with @FabianMeiswinkel 's point that a 2 minute latency is equivalent to availability loss for almost all applications and maybe we should just fail at that point.

> But IMO an exponential backoff - strating with maybe 200ms and having an upper limti of somewehre around2-3 seconds back-off would make more sense.

It would make sense if the failing requests are causing significant impact, but an exponential backoff would also cause medium-duration topology updates to result in much longer latency (if the update takes 30s but our exponential backoff is already up to 60s by that point, or something like that). It'd be nice if we had some data on the expected duration of these updates. Not just a ceiling, but a distribution of sorts.

**Comment by analogrelay** [comment]
I re-spun the failed job, but it looks like it failed _after_ completing the tests successfully. They just took long enough that the clean-up process took it over the timeout. Given that, I'm OK with you merging without getting fully green on the weekly tests, as long as the required checks are good.

Maybe we need to increase the run timeout, or fan out a bit more so the runs aren't as long.

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md (hunk 8 above)
Comment (tvaron3) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md:60]: Makes sense agree. Here is the issue https://github.com/Azure/azure-sdk-for-rust/issues/4620 It includes gathering some results on the amount of time for a partitionto become healthy again during the topology changes.https://github.com/Azure/azure-sdk-for-rust/issues/4620
